src/Raven.Server/Documents/Handlers/Processors/Documents/AbstractDocumentHandlerProcessorForGet.cs

```diff
@@ -493,4 +503,228 @@ protected sealed class StartsWithParams
 
         public string StartAfterId { get; set; }
     }
+
+    public sealed class QueryStringParameters : AbstractQueryStringParameters
+    {
+        public bool MetadataOnly;
+
+        public StringValues IncludePaths;
+
+        public List<ReadOnlyMemory<char>> Ids;
+
+        public StringValues Counters;
+
+        public List<ReadOnlyMemory<char>> Revisions;
+
+        public ReadOnlyMemory<char>? RevisionsBefore;
+
+        public List<ReadOnlyMemory<char>> TimeSeries;
+
+        public bool TimeSeriesHasAllTimeSeries;
+
+        public List<ReadOnlyMemory<char>> TimeSeriesTimes;
+
+        public bool TimeSeriesTimesHasAllTimeSeries;
+
+        public List<ReadOnlyMemory<char>> TimeSeriesCounts;
+
+        public bool TimeSeriesCountsHasAllTimeSeries;
+
+        public StringValues From;
+
+        public StringValues To;
+
+        public StringValues TimeTypes;
+
+        public StringValues TimeValues;
+
+        public StringValues TimeUnits;
+
+        public StringValues CountTypes;
+
+        public StringValues CountValues;
+
+        public StringValues CompareExchange;
+
+        public TransactionMode TxMode;
+
+        private readonly bool _isGet;
+
+        private QueryStringParameters([NotNull] HttpRequest httpRequest)
+            : base(httpRequest)
+        {
+            _isGet = httpRequest.Method == HttpMethods.Get;
+        }
+
+        protected override void OnFinalize()
+        {
+            if (AnyStringValues() == false)
+                return;
+
+            IncludePaths = ConvertToStringValues("include");
+            Counters = ConvertToStringValues("counter");
```

Comment:
Why are we performing a dictionary check if this could be instantiated once.

Reply:
This is using generic mechanism that can be used in other places, in general Counters, IncludePaths, etc are still StringValues. That would require major rewriting to handle those. Primary advantage of changes in this PR is 'id' parsing and using ReadOnlyMemory<char> for id to avoid allocating string back and forth. Other stuff I've left as it was... for now.

Reply:
Then we should add an issue with all the knowledge you acquired during this optimization to tackle that in the future.